### ISSUE TITLE
interacting with $CAGEBREAK_SOCKET

### ISSUE THREAD

**Issue by exit666** [project-repo/cagebreak#29]
It's not an issue I guess, I just couldn't find any information on how to interact with the socket. 
```
nc -u $CAGEBREAK_SOCKET 
```
I get `forward host lookup failed`. So how do I interact with this socket? Sorry for a noob question, but I would appreciate some examples.  If that's inappropriate place to discuss but someone wants to help me -- my discord is exitone hashtag 9183.

**Comment by project-repo** [comment]
Sure, no problem, questions are always welcome!

One problem might be that the syntax for the `nc` command is:
```
nc -U $CAGEBREAK_SOCKET
```
i.e. with an uppercase "U". Can you try that?

If that works, you should get a connection to the cagebreak socket and be able to type commands as
in the configuration file (see example file [here](https://github.com/project-repo/cagebreak/blob/master/examples/config)).
Does that help?

Or what kind of examples would you like to see more specifically?

Cheers
project-repo

**Comment by exit666** [comment]
Thanks for reply!
The problem was my version of netcat weirdly didn't have -U option. Had to swap it for openbsd-netcat version in gentoo's repository. I wanted examples of scripts because I didn't know how to connect. Thanks, all works now!

**Comment by sodface** [comment]
I just wanted to post a few notes on interacting with the IPC socket using busybox on Alpine Linux and rather than open a new issue, I thought added a comment here would be ok?

The example scripts in this repo have dependencies on:
- bash
- openbsd-netcat (Alpine package is named [netcat-openbsd](https://pkgs.alpinelinux.org/package/edge/main/armhf/netcat-openbsd))
- jq

All of these are available in the Alpine package repos but they are not included in a default Alpine installation, which uses busybox versions of many common linux utilities instead, including the ash shell and nc.

I was able to get the equivalent functionality using just busybox by making the following changes:

https://github.com/project-repo/cagebreak/blob/bea18c7a6f0f709251d6e7670eed5d3d9ef1b02b/example_scripts/cb_script_header.sh#L9

Change to:
```
busybox nc local:"${CAGEBREAK_SOCKET}" < "${named_pipe_send}" > "${named_pipe_recv}"&
```

If you keep the above script header in a separate file and source it in your scripts, change:

https://github.com/project-repo/cagebreak/blob/bea18c7a6f0f709251d6e7670eed5d3d9ef1b02b/example_scripts/focus_follows_cursor.sh#L10ca

To:

```
source "$(dirname "$0")/cb_script_header.sh"
```

Parsing the json events without jq isn't as straight forward but I didn't really want to work with json anyway and looked for a way to convert the key/value pairs into shell variables. [This question on Stackoverflow ](https://stackoverflow.com/questions/25378013/how-to-convert-a-json-object-to-key-value-format-in-jq) provided some answers on the key/value conversion step using jq.

With a cagebreak event for example:
```
~ $ printf "${event:6}\n" | jq -r 'to_entries|map("\(.key)=\(.value|tostring)")|.[]'
event_name=cursor_switch_tile
old_output=HDMI-A-1
old_output_id=1
old_tile=1
new_output=HDMI-A-2
new_output_id=2
new_tile=2
```

Then I came up with a busybox awk version that does mostly the same thing, except it wraps all the values in quotes:

```
~ $ printf "${event:7:-1}" | busybox awk -F":" -v RS=',"' '{gsub(/\"/,"");print $1"=\""$2"\""}'
event_name="cursor_switch_tile"
old_output="HDMI-A-1"
old_output_id="1"
old_tile="1"
new_output="HDMI-A-2"
new_output_id="2"
new_tile="2"
```

Out of curiosity, I looped each version 1000 times and timed it. The awk version is much faster:

busybox awk:
```
real	0m 0.50s
user	0m 0.46s
sys	0m 0.13s
```
jq
```
real	0m 28.03s
user	0m 26.93s
sys	0m 1.18s
```

I realize the awk version is probably more fragile but so far it's working ok.
